File: 2015/Day17/main.dart
import 'dart:convert';
import 'dart:io';

Iterable<List<int>> combinations(List<int> elements, int r) sync* {
  if (r == 0) {
    yield [];
  } else {
    for (int i = 0; i <= elements.length - r; i++) {
      for (var combo in combinations(elements.sublist(i + 1), r - 1)) {
        yield [elements[i], ...combo];
      }
    }
  }
}

int countEggnogCombinations(List<int> containers, int target) {
  int count = 0;
  for (int r = 1; r <= containers.length; r++) {
    for (var combo in combinations(containers, r)) {
      if (combo.reduce((a, b) => a + b) == target) {
        count++;
      }
    }
  }
  return count;
}

Map<String, int> findMinContainersAndCombinations(
    List<int> containers, int target) {
  for (int r = 1; r <= containers.length; r++) {
    int count = 0;
    for (var combo in combinations(containers, r)) {
      if (combo.reduce((a, b) => a + b) == target) {
        count++;
      }
    }
    if (count > 0) {
      return {'minContainers': r, 'combinations': count};
    }
  }
  return {'minContainers': 0, 'combinations': 0};
}

void main() async {
  final file = File('input.txt');
  List<int> containers = await file
      .openRead()
      .transform(utf8.decoder)
      .transform(LineSplitter())
      .map(int.parse)
      .toList();

  int targetVolume = 150;

  print('Part 1: ${countEggnogCombinations(containers, targetVolume)}');
  print(
      'Part 2: ${findMinContainersAndCombinations(containers, targetVolume)['combinations']}');
}

File: 2015/Day17/main.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <numeric>
#include <string>
#include <vector>

void combinations(const std::vector<int> &elements, int r, int start,
                  std::vector<int> &current,
                  std::vector<std::vector<int>> &result) {
  if (r == 0) {
    result.push_back(current);
    return;
  }

  for (int i = start; i <= elements.size() - r; ++i) {
    current.push_back(elements[i]);
    combinations(elements, r - 1, i + 1, current, result);
    current.pop_back();
  }
}

int count_eggnog_combinations(const std::vector<int> &containers, int target) {
  int count = 0;
  for (int r = 1; r <= containers.size(); ++r) {
    std::vector<std::vector<int>> result;
    std::vector<int> current;
    combinations(containers, r, 0, current, result);
    for (const auto &combo : result) {
      if (std::accumulate(combo.begin(), combo.end(), 0) == target) {
        count++;
      }
    }
  }
  return count;
}

std::map<std::string, int>
find_min_containers_and_combinations(const std::vector<int> &containers,
                                     int target) {
  for (int r = 1; r <= containers.size(); ++r) {
    int count = 0;
    std::vector<std::vector<int>> result;
    std::vector<int> current;
    combinations(containers, r, 0, current, result);
    for (const auto &combo : result) {
      if (std::accumulate(combo.begin(), combo.end(), 0) == target) {
        count++;
      }
    }
    if (count > 0) {
      return {{"min_containers", r}, {"combinations", count}};
    }
  }
  return {{"min_containers", 0}, {"combinations", 0}};
}

int main() {
  std::ifstream file("input.txt");
  std::vector<int> containers;
  std::string line;
  while (std::getline(file, line)) {
    containers.push_back(std::stoi(line));
  }

  int target_volume = 150;

  std::cout << "Part 1: "
            << count_eggnog_combinations(containers, target_volume)
            << std::endl;
  std::cout << "Part 2: "
            << find_min_containers_and_combinations(
                   containers, target_volume)["combinations"]
            << std::endl;
}
